scripts/install-protoc.sh

```diff
@@ -39,53 +35,66 @@ source "$(dirname $0)/common.sh"
+    "x86_64")   ARCH="x86_64";;
+    "aarch64")  ARCH="aarch_64";;
+    "arm64")    ARCH="aarch_64";;
+    *)          die "Install unsuccessful. Hardware platform not supported by installer: $1";;
+  esac
+  case "${2}" in
+    "Darwin") OS="osx";;
+    "Linux")  OS="linux";;
+    *)        die "Install unsuccessful. OS not supported by installer: $2";;
+  esac
+
+  # Check if the protoc binary with the right version is already installed.
+  if [[ -f "${INSTALL_PATH}/bin/protoc" ]]; then
+    if [[ "$("${INSTALL_PATH}/bin/protoc" --version)" == "libprotoc ${PROTOC_VERSION}" ]]; then
+      echo "protoc version ${PROTOC_VERSION} is already installed in ${INSTALL_PATH}."
+      return
+    fi
+  fi
+
+  DOWNLOAD_URL="https://github.com/protocolbuffers/protobuf/releases/download/v${PROTOC_VERSION}/protoc-${PROTOC_VERSION}-${OS}-${ARCH}.zip"
   # -L follows redirects.
   # -O writes output to a file.
   curl -LO "${DOWNLOAD_URL}"
-  unzip "protoc-${PROTOC_VERSION}-$2-$1.zip" -d "${INSTALL_PATH}"
-  rm "protoc-${PROTOC_VERSION}-$2-$1.zip"
-  rm "${INSTALL_PATH}/readme.txt"
+  # Unzip the downloaded file and except readme.txt.
+  # The file structure should look like:
+  # INSTALL_PATH
+  # ├── bin
+  # │   └── protoc
+  # └── include
+  #     └── other files...
+  unzip "protoc-${PROTOC_VERSION}-${OS}-${ARCH}.zip" -d "${INSTALL_PATH}" -x "readme.txt"
+  rm "protoc-${PROTOC_VERSION}-${OS}-${ARCH}.zip"
+  # Make the protoc binary executable. ¯\_(ツ)_/¯  crazy, right?
+  chmod +x "${INSTALL_PATH}/bin/protoc"
 }
 
 main() {
-  # Check if protoc is already available.
-  if command -v protoc &> /dev/null; then
-    if INSTALL_VERSION=$(protoc --version | cut -d' ' -f2 2>/dev/null); then
-      if [ "$INSTALL_VERSION" = "$PROTOC_VERSION" ]; then
-        echo "protoc version $PROTOC_VERSION is already installed."
-        return
-      else
-        die "Existing protoc version ($INSTALL_VERSION) differs. Kindly make sure you have $PROTOC_VERSION installed."
-      fi
-    fi
+  if [[ "$#" -ne 1 ]]; then
+    die "Usage: $0 INSTALL_PATH"
   fi
 
-  # Detect the architecture
-  case "$(uname -m)" in
-    "x86_64") ARCH="x86_64";;
-    "aarch64") ARCH="aarch_64";;
-    "arm64") ARCH="aarch_64";;
-  *) die "Unsupported architecture. Please consider manual installation from "\
-          "https://github.com/protocolbuffers/protobuf/releases/ and add to PATH."
-  esac
-
-  # Detect the Operating System
-  case "$(uname -s)" in
-    "Darwin") download_binary $ARCH "osx";;
-    "Linux") download_binary $ARCH "linux";;
-  *) die "Unsupported OS. Please consider manual installation from "\
-          "https://github.com/protocolbuffers/protobuf/releases/ and add to PATH" ;;
-  esac
+  if [[ ! -d "${INSTALL_PATH}" ]]; then
+    die "INSTALL_PATH (${INSTALL_PATH}) does not exist."
+  fi
+  echo "Installing protoc version $PROTOC_VERSION to ${INSTALL_PATH}."
+  # Detect the hardware platform.
+  ARCH="$(uname -m)"
+  # Detect the Operating System.
+  OS="$(uname -s)"
```

Comment:
I'd probably delete these vars and do these things in `download_binary` instead.  Note that you're setting `ARCH` twice to different things, so that's a little confusing, and you're passing these as positional parameters, so separating their usage from their definition.

Reply:
Yes you are right. I dont know why we needed a function to be defined like this. I've moved everything into main() now. Looks cleaner this way.